### ISSUE TITLE
Tai Tham uses subjoined consonants after right-side vowels

### ISSUE THREAD

**Issue by roozbehp** [harfbuzz/harfbuzz#170]
Basically, the regular expression used in Universal Shaping engine doesn't allow subjoined consonants after right-side vowels (visually hanging under them), while Tai Tham seems to be using them.

Specific examples should be available from Martin and Patrick, CC-ed.

/cc @mhosken @pchew-change

**Comment by behdad** [comment]
Shouldn't this be solved at the USE level first?

**Comment by roozbehp** [comment]
Not necessarily. We cannot wait for USE. We know we need to allow the sequences.

**Comment by behdad** [comment]
But I rather not diverge from it.  We don't need to wait for USE spec to be updated, but at least you and Andrew to decide how you want to address it.

**Comment by pchew-change** [comment]
some examples include:

ᩈᩣ᩠ᨾ (ᩈ | ᩣ | ᩠ | ᨾ) U+1A48 U+1A63 U+1A60 U+1A3E
(‘three’)

ᩃ᩶ᩣ᩠ᨶ (ᩃ | ᩣ | ᩠ | ᨶ) U+1A43 U+1A76 U+1A63 U+1A60 U+1A36
(‘myriad, ten-thousand’)

ᨷ᩶ᩤ᩠ᨶ (ᨷ | ᩶ | ᩤ | ᩠ | ᨶ) U+1A37 U+1A76 U+1A64 U+1A60 U+1A36
(‘home, village’)

ᨩ᩶ᩣ᩠ᨦ (ᨩ | ᩶ | ᩣ |  ᩠ | ᨦ) U+1A29 U+1A76 U+1A63 U+1A60 U+1A26
(‘elephant’)

ᩃᩖᩣ᩠ᨿ (ᩃ | ᩖ | ᩣ |  ᩠ | ᨿ) U+1A43 U+1A56 U+1A63 U+1A60 U+1A3F
(‘many’)

ᨠᩣ᩠ᩁ (ᨠ | ᩣ |  ᩠ | ᩁ) U+1A20 U+1A63 U+1A60 U+1A41
(‘the process of…, the activity of…, an enactment of…, the state of…’ )

ᩈᩣ᩠ᩅ (ᩈ | ᩣ |  ᩠ | ᩅ) U+1A48 U+1A63 U+1A60 U+1A45
(‘young woman’)

Please let me know if you need more.

**Comment by pchew-change** [comment]
Rendered versions in the attached png.
![vowel_right_sakot_c](https://cloud.githubusercontent.com/assets/1437602/11482419/8178b872-9757-11e5-9e60-985cd190e9e0.png)

**Comment by brawer** [comment]
Has anyone heard from Andrew?

**Comment by roozbehp** [comment]
@brawer It's probably on his plate with some lower priority, but I think you should contact him directly to get it higher. Also note that we don't need to wait for Andrew. We have fixed things in HarfBuzz in contrast with the OpenType spec before.

**Comment by behdad** [comment]
> We have fixed things in HarfBuzz in contrast with the OpenType spec before.

But we need a proposed resolution.

**Comment by brawer** [comment]
@pchew-change, to resolve this, could you make a copy of the [USE intro document](https://www.microsoft.com/typography/OpenTypeDev/USE/intro.htm), with (clearly marked) modifications to support Tai Tham? Then we’d have a concrete proposal to discuss.

**Comment by Richard57** [comment]
At Roozbeh's request I drafted a list of test examples last year (2015) but was never told whether it was in a suitable form - it's available (with revisions) at http://www.wrdingham.co.uk/lanna/renderer_test.htm .

The title of this issue is misleading - almost any vowel can be followed by a subjoined consonant.  The exceptions are SIGN A (which appears to be the visarga, though /ʔ/ rather than /h/) and SIGN OA below (geometry, though perhaps Patrick can find some examples).  The name 'SAKOT', used for the invisible stacker, means 'coda consonant'.  I wouldn't rule out a silent final consonant under SIGN AI, either.

**Comment by pchew-change** [comment]
Hi, Richard !

I'm getting a 404 error when I add an 'o' to your domain and a "
www.wrdingham.co.uk’s server DNS address could not be found." without it. =(

I'll see if I can try and hunt for more examples...

-------------------------------------------------------------------
Patrick Chew
Internationalization Manager, Change.org
[ 🌐Globalization | Ìńťẽřñățıöņåłįżȁṯỉơṋ | Localisation | Translation→翻譯]
pchew@change.org | 415.689.9768
Twitter: @PChewWork <http://twitter.com/PChewWork>
Skype: pchew.change

  Change.org <http://change.org/>
  Follow us on: LinkedIn <http://www.linkedin.com/company/459396> | Twitter
<http://twitter.com/changeorgs> | Facebook
<http://www.facebook.com/Change.Org>

**Comment by Richard57** [comment]
On Thu, 10 Nov 2016 15:51:39 -0800
pchew-change <notifications@github.com> wrote:

> I'm getting a 404 error when I add an 'o' to your domain and a "
> www.wrdingham.co.uk’s server DNS address could not be found." without
> it. =(

The server must have gone down.  I can't refresh my downloads of the
page.  Someone else got the 'o' - I was way too late registering a
domain.

> I'll see if I can try and hunt for more examples...

I trust you mean for rare VC combinations.  About half the words in the
local languages have CVC orthographic syllables, even in Northern Thai.

Richard.

**Comment by Richard57** [comment]
The registration of my domain name has been suspended; it looks as though it won't be sorted out until Monday at the earliest.  In the mean time, I've created a synonymous subdomain tied to the web host, so the test set is accessible as http://richardwordingham.me.pn/lanna/renderer_test.htm .

**Comment by Richard57** [comment]
I've quantified the damage done to rendering Northern Thai in Tai Tham by dropping the SEA shaper from HarfBuzz at http://wrdingham.co.uk/lanna/random_test.htm .  I sampled the lexicon, but the results seem surprisingly representative - the bias to Pali may have been balanced out by expressive 4-syllable words.  If one merely wants to convey a plain text message using the script then:

46% of words are undamaged by the engine.
37% of words can be expressed, more or less, by substituting a homographic (or near homographic) anagram.

Most of the remaining 17% could be recovered by treating U+1A63 TAI THAM VOWEL SIGN AA and U+1A64 TAI THAM VOWEL SIGN TALL AA as though they had gc=Lo.  Equivalently, one could just override their Indic_Syllabic_Category (isc) to make them 'Consonant'.  I've long had a suspicion that U+1A63 is really gc=Lo, just like U+0E32 THAI CHARACTER SARA AA.  The one possible downside I can see is that the ligature NAA from NA + AA could not be formed until syllable boundaries were dissolved, but at present an intervening tone mark would also wrongly prevent ligation at that change.

For the northern languages, it would be helpful to also make U+1A6D TAI THAM VOWEL SIGN OY a consonant.  That would enable the rendering of the common word ᨶ᩶ᩭ  <U+1A36 TAI THAM LETTER NA, U+1A76 TAI THAM SIGN TONE-2, U+1A6D> 'little'.

These constitute a 'quick fix' that may be reversed when a new regex is defined for USE.  Of course, another 'quick fix' might be to re-enable the SEA shaper for Tai Tham; I don't know how 'quick' that would be.

Correcting the isc of U+1A5B TAI THAM CONSONANT SIGN HIGH RATHA OR LOW PA to Consonant_Subjoined now would mop up the other 2 cases from my sample of 410 words.  Oddly enough, I didn't sample any problems caused by the isc of U+1A56 TAI THAM CONSONANT SIGN MEDIAL LA being Consonant_Medial; they would be eliminated by changing its isc to Consonant_Subjoined.  (There are examples of the problem in my test set.)

I believe the isc of U+1A55 TAI THAM CONSONANT SIGN MEDIAL RA should also be overridden to Consonant_Subjoined; my only concern is that this might interfere with the pref feature's tagging it as a medial consonant to be re-ordered to pre-base position.

A general revision of USE may make overriding the isc of U+1A55 and U+1A56 redundant; there are other scripts where a medial consonant may be immediately followed by a subjoined consonant.

**Comment by Richard57** [comment]
There is a false perception that "... Tai Tham polysyllabic clusters. This is more of a philosophical debate on orthographic clusters and reading order" (e.g. its expression at https://github.com/behdad/harfbuzz/issues/376 .)  The sequences <U+1A69 TAI THAM VOWEL SIGN U, U+1A60 TAI THAM SIGN SAKOT, U+1A3F TAI THAM LETTER LOW YA>, as in ᩉᩩ᩠ᨿ, and  <U+1A60 TAI THAM SIGN SAKOT, U+1A3F TAI THAM LETTER LOW YA, U+1A69 TAI THAM VOWEL SIGN U>, as in ᩉ᩠ᨿᩩᨠ, are frequently written differently in Northern Thai.

**Comment by behdad** [comment]
cc @dscorbett 

On the Unicode mailing list Andrew Glass said the following.  I think we should implement it.

Here is the essence of the initial changes needed to support CV+C. Open to feedback.

 

Create new SAKOT class
SAKOT (Sk) based on UISC = Invisible_Stacker
Reduced HALANT class
Now only HALANT (H) based on UISC = Virama
Updated Standard cluster mode
[< R | CS >] < B | GB > [VS] (CMAbv)* (CMBlw)* (< < H | Sk > B | SUB > [VS] (CMAbv)* (CMBlw)*)* [MPre] [MAbv] [MBlw] [MPst] (VPre)* (VAbv)* (VBlw)* (VPst)* (VMPre)* (VMAbv)* (VMBlw)* (VMPst)* (Sk B)* (FAbv)* (FBlw)* (FPst)* [FM]


The only required component of a standard cluster is a BASE or BASE_OTHER. A cluster may optionally begin with a REPH or CONS_WITH_STACKER. A BASE or BASE_OTHER may be followed immediately by a VARIATION_SELECTOR and/or multiple CONS_MOD characters in the order CONS_MOD_ABOVE CONS_MOD_BELOW. Multiple sequences of a HALANT BASE or SAKOT BASE with optional VARIATION_SELECTOR or optional CONS_MOD can occur. The sequence can continue with zero or one CONS_MED for each cardinal position (Pre, Above, Below, Post); zero to many VOWEL characters in each cardinal position; zero to many VOWEL_MODs in each cardinal position; zero to many sequences of SAKOT BASE; zero to many CONS_FINALs in each of Above, Below, and Post; and lastly, an optional FINAL_MOD.

 

Updated Halant-terminated cluster
[< R | CS >] < B | GB > [VS] (CMAbv)* (CMBlw)* (< < H | Sk > B | SUB > [VS] (CMAbv)* (CMBlw)*)* < H | Sk >
 

This is similar to the Standard cluster but terminates in a final HALANT or SAKOT after a BASE, BASE_OTHER, or CONS_MOD. When such a HALANT or SAKOT it will form a cluster. When any character other than a BASE or BASE_OTHER follows the HALANT or SAKOT there will be a cluster break between the HALANT or SAKOT and the following character. Multiple sequences of a HALANT BASE or SAKOT BASE with optional VARIATION_SELECTOR or optional CONS_MOD can occur. A CONS_SUBJ is equivalent to the sequence HALANT BASE.

 

New Sakot-terminated cluster
[< R | CS >] < B | GB > [VS] (CMAbv)* (CMBlw)* (< < H | Sk > B | SUB > [VS] (CMAbv)* (CMBlw)*)*

    [MPre] [MAbv] [MBlw] [MPst]

    (VPre)* (VAbv)* (VBlw)* (VPst)*

    (VMPre)* (VMAbv)* (VMBlw)* (VMPst)*

    (Sk B [VS] (CMAbv)* (CMBlw)*)* Sk

 

This is similar to the Standard cluster but terminates in a final SAKOT after a VOWEL or VOWEL_MOD. When such a SAKOT follows a VOWEL or VOWEL_MOD it will form a cluster. When any character other than a BASE or BASE_OTHER follows this SAKOT there will be a cluster break between the SAKOT and the following character. Multiple sequences of a SAKOT BASE with optional VARIATION_SELECTOR or optional CONS_MOD can occur. A CONS_SUBJ is equivalent to the sequence HALANT BASE.


This would allow a consonant to follow a vowel when joined with a Sakot. It would support multiple final consonants. It would not support polysyllabic chaining of CV+CV+CV etc.

 

Cheers,

**Comment by mhosken** [comment]
One problem here is that a tone mark is VMAbv and may be followed by a Sk B. But due to the bug in normalization, that will have been reordered to Sk VMAbv B! I guess it's back to the UTC for that one.

**Comment by Richard57** [comment]
Of course, that's not a problem for a truly Unicode-compliant process.  I believe HarfBuzz still has the code that chooses an alternative canonical form for Tai Tham, by effectively changing the combining class of SAKOT to something greater than 230.  So doing solved the problem with the SEA shaper.

What's the UTC got to do with the USE?

**Comment by dscorbett** [comment]
Defining Sk as UISC = Invisible_Stacker is too broad. It introduces ambiguity to scripts where there is no need to subjoin consonants after vowels.

What is the purpose of the sakot-terminated cluster? It is my understanding that sakot has no independent existence and should always precede a consonant. Is it just to support incomplete syllables while typing?

Is C V Sk C intended to be rendered potentially the same as C Sk C V, with the encoding based on the pronunciation; or is it only to be used when the final consonant visually comes “after” (i.e. below or to the right of) the vowel sign?

[L2/18-332 “Tai Tham ad hoc meeting report”](https://www.unicode.org/L2/L2018/18332-tai-tham-ad-hoc-report.pdf) recommends:
>* A written structure <C1 base, C2 sign, V sign, C3 base> shall not be represented phonetically with <C1 letter, C2 sign, V sign, C3 sign>. (Note a consonant sign can be either an atomic character or a <stacker, consonant letter> sequence.)
> * To represent a cluster, regardless of the phonetic order CCV or CVC, a consonant sign should always be encoded before the vowel sign, unless the vowel sign has inline advance and is apparently followed by the consonant sign.

@lianghai, @mhosken, @patchew, @roozbehp, and @xadxura, you were all at that meeting: can you comment on the current state of the Tai Tham encoding discussion? Will Tai Tham text use a phonetic or a “visual-motivated” order, and how does that affect sakot placement?

**Comment by Richard57** [comment]
My understanding that final SAKOT is for incompletely typed or incompletely deleted clusters.

The original intention, not rescinded by the UTC, is that Tai Tham is encoded in phonetic order where discernible, and so the basic difference between C V Sk C and C Sk V C.  However, there is often a visible difference.  Discussions have assumed that there is a difference when V is postscript vowel.  This is true for U+1A63, usually true for U+1A64, and can be false for U+1A6D.  I've never seen a subscript consonant follow U+1A61.

There seems to be no difference for prescript vowels.

For spacing subscript consonants, there is often a small difference in the placement of a superscript vowel.  It is further to the left for CVC than for CCV.  I originally thought this was an artefact of pre-Unicode fonts, but I have also found the difference in handwriting.  The VC-CV pairs /uy/ v. /yu/ and /uːy/ v. /yuː/ are generally distinct in Northern Thai, not Tai Khuen.  In Northern Thailand, the vowel in the first case appears to the bottom left.  The Northern Thai dictionary of Palm Leaf Manuscripts similarly distinguishes the VC-CV pair /u:p/ v. /pu:/ - the examples are ᨡᩕᩡᩋᩪ᩠ᨷᨠᩯ᩠᩶ᩅ 'the mouth of the Lord Buddha' on p55 and ᩅᩤ᩠ᨯᩁᩪ᩠ᨷ (normal meaning 'to draw a picture') on p643 - this contrast is unusual, but as the writer, 'it's up to you'!

Some subscript consonants are optionally spacing, so the existence of a difference may depend on the font!

Note that for prescript vowels and a spacing subscript, CVTC and CCVT may contrast in the placement of the tone mark.

In view of these contrasts, I strongly urge that that abominable recommendation be ignored.

**Comment by mhosken** [comment]
The sakot terminated cluster is a rarity but can crop up in things like 1A3B 1A66 1A76 1A60 1A36 1A6C 1A26 (pinong). The good news about this approach taken in the USE is that it allows for the variety of spelling orders currently in existence.

If harfbuzz can get around the canonical ordering problem (e.g. the above string would be normalized to 1A3B 1A66 **1A60 1A76** 1A36 1A6C 1A26). Then I suggest we code it up and throw some real data and fonts at it and see how it stands up. Tai Tham is complex enough script that predicting the problems before we start would result in us never starting.

As to the desire that many have expressed for a more regularised and visually motivated encoding order that pushes the ambiguity away from the storage order back to the visual, where it belongs, it is a difficult problem, as Richard points out. I do struggle a bit with the idea that CVTC and CCVT would result in a positional difference for the tone. Nothing in the script relies on that. So at best it is a typographical nuance. But I do concur on the /uy/ vs /yu/ and therefore would not be surprised by a contrast between /u:p/ and /pu:/. This just means we have to think harder. There are many wins to regularising the order away from linguistic purity in terms of making things easier to type. If you have a fixed encoding order, you can allow for a variety of typing orders. But if you have a variety of encoding orders you have to have a fixed typing order.

**Comment by dscorbett** [comment]
The proposed new standard cluster ends in (Sk B)* (FAbv)* (FBlw)* (FPst)* [FM]. The proposed new sakot-terminated cluster ends in (Sk B [VS] (CMAbv)* (CMBlw)\*)* Sk. This is inconsistent. Should CONS_MOD be allowed after Sk B or not?

**Comment by Richard57** [comment]
"1A3B 1A66 1A76 1A60 1A36 1A6C 1A26 (pinong)" should be rendered as two stacks, the last being just one Unicode character.   The U+1A60 marks a phonetic syllable boundary, but not an orthographic syllable boundary.  The U+1A36 is either the next to bottom character of the first stack, or wraps the U+1A6C.  What I would expect to see is 3 stacks: <1A3B 1A66 1A76 1A60 1A36> <25CC* 1A6C> <1A26>.  The U+25CC is localised (feature locl) on Windows, but not in HarfBuzz.

I was wondering why CVTC and CCVT should be different with a prescript vowel and a spacing subscript consonant.    A possible cause is the positioning of the tone mark in the Thai script.  I'm not sure that that should be called a 'typographical nuance', though I wouldn't trust its placement in hurried writing.  Siamese dictionaries rely on this to avoid saying whether the syllable is CVC or CCV.  Without a tone mark, the RID uses a carefully placed phinthu in the phonetic respelling.

Talking of complexity, the USE order certainly increases it.  A lot of fonts have managed to avoid mark attachment classes because interacting marks are adjacent in unchained syllables.  The USE jumps all over the place - subscript consonant, vowel before, vowel above, vowel below, vowel after, tone mark.  A lot of fonts that don't use the renderer's Indic rearrangement fail with subscript + vowel before.  Toggling the relationship between Indic positional class and Vabv and Vblw, as has been done elsewhere, would help a lot of limited fonts!  I've been treating spacing marks as bases; this will have to change if the UTC hauls up the white flag.  I think I'd also need to use the innovation of the mark attachment *sets* so that when chaining is allowed, U+1A63 will block subsequent subscript consonants and then subsequent non-spacing vowels.

**Comment by Richard57** [comment]
Allowing CONS_MOD after a subscript coda consonant makes sense.  It would permit <SAKOT, GREAT SA, KARAN>, which makes sense as the apocopation of a loanword from Pali ending in -ssa.  (The KARAN removes the implicit vowel, and the second /s/ then drops.)  Fiddling the category would then enable us to make U+1A5B TAI THAM CONSONANT SIGN HIGH RATHA OR LOW PA a consonant modifier (I think it always follows a consonant, which it geminates and possibly aspirates).  We still don't enable <SAKOT, GREAT SA, RA HAAM>; this is a chained syllable, as RA HAAM is a pure killer, and so can now silence the first consonant in a vowelless two syllable consonant stack, as in the borrowings of 'serve' (think tennis).

**Comment by dscorbett** [comment]
There’s another problem with the proposed cluster definition: it allows both B M Sk B and B Sk B M, and it allows B Sk B M V, but not B M Sk B V. The intent is that a subjoined consonant can follow a vowel sign, but nothing checks that there is at least one vowel sign; that makes a subjoined consonant able to immediately follow a medial consonant _unless_ a vowel sign follows it, which is inconsistent.

One way to fix this is to change the regex from (essentially) B ((H | Sk) B)\* M\* V\* (Sk B)\* to B ((H | Sk) B)\* M\* (V+ (Sk B)\*)?. That is, only allow the post-vowel subjoined consonant if there actually is a vowel.

**Comment by Richard57** [comment]
Two quick points on this.  Firstly, the USE should not have the exemplifiable concept of Tai Tham medial consonants, and fixing that problem is a simple data table tweak.  (Remember that Sanskrit has both grv and gvr clusters.)  Secondly, in that proposed regex, 'V' must also include tone marks, as needed for ᨯ᩠ᩅ᩠᩶ᨿ /duai/ 'also, together' <LETTER DA, SAKOT, LETTER WA, TONE-2, SAKOT, LOW YA>.   Once one imposes requirements on the order of the constituents of 'V*', I am not sure that we still have a practicable 'deterministic regex', whatever the esoteric meaning of that ambiguous expression is.

As a theoretical aside, I fear that the deterministic expression of the intersection of the languages a*b*c* and {a|b|c)+ is something like (a+b*c*|b+c*|c+), which gets unwieldy with 8 types of constituent (though only 5 are non-empty when restricted to Tai Tham).

**Comment by mhosken** [comment]
> "1A3B 1A66 1A76 1A60 1A36 1A6C 1A26 (pinong)" should be rendered as two stacks, the last being just one Unicode character. 

I agree for modern spelling conventions. But this example, as is, is found in "TuaMuang" by Kasem Siriwatpiriyaa, p159. People should be able to render pretty much whatever crazy spelling they want, within some level of reason, if the font so desires to support it. The question is how to ensure a suitable level of contrast in spelling while providing an easy to use encoding. As in, if we can standardise the encoding visually then we get all kinds of keyboarding goodness for users.

**Comment by mhosken** [comment]
> There’s another problem with the proposed cluster definition: it allows both B M Sk B and B Sk B M, and it allows B Sk B M V, but not B M Sk B V. The intent is that a subjoined consonant can follow a vowel sign, but nothing checks that there is at least one vowel sign; that makes a subjoined consonant able to immediately follow a medial consonant _unless_ a vowel sign follows it, which is inconsistent.

Counter to Richard, there are special characters that act as medials (U+1A55..U+1A56). I sympathise with his desire to not encourage medials too much due to the multi-valent use of some characters, like SAKOT WA. But either way, I can't think of a sequence that would fall fowl of B M Sk B V, but Richard may, likewise for B Sk B M V. Hmm /ontarai/ 'danger' may fit the bill <LETTER A, VOWEL O, LETTER NA, SAKOT, LETTER TA, MEDIAL RA, VOWEL AI>

**Comment by Richard57** [comment]
I said, "Firstly, the USE should not have the exemplifiable concept of Tai Tham medial consonants, and fixing that problem is a simple data table tweak."  The tweak is to reclassify U+1A55 and U+1A56 as consonant_subscript in the data tables used for the USE.  Then, as far as the USE is concerned, Tai Tham will have no medial consonants.

For B M Sk B V, we have the immediate examples of ᨠᩖ᩠ᩅᩫ /kua/ 'to fear' <HIGH KA, MEDIAL LA, SAKOT, WA, SIGN O> and ᨣᩕ᩠ᩅᩫ  /kʰua/ 'household' <LOW KA, MEDIAL RA, SAKOT WA, SIGN O> and the irrefutable ᩉᩖ᩠ᩅᩣ /lwaː/ 'iron' <HIGH HA, MEDIAL LA, SAKOT, WA, SIGN AA>.  There's even the Tai Khuen ᨲᩕ᩠ᩃᩬᨯ /tʰălɔːt/ 'for ever' <HIGH TA, MEDIAL RA, SAKOT LA, SIGN OA BELOW, DA>.  (Or was there a Northern Thai typesetter at work?)  In the last example, the stack bent below the DA, as is usual for modern Tai Khuen (and quite common in Northern Thailand).  That should just be the font-creator's problem.

**Comment by Richard57** [comment]
> But this example, as is, is found in "TuaMuang" by Kasem Siriwatpiriyaa, p159<
Picture, please.  I can't imagine what an old 'sakot terminated stack' would look like.

**Comment by Richard57** [comment]
For B Sk B M V, which the USE has never challenged, there is the impeccable Pali word _indriya_ 'sense' and its numerous compounds, e.g.  [ᩈᨶ᩠ᨲᩥᨶ᩠ᨴᩕᩥᨿ](https://en.wiktionary.org/wiki/%E1%A9%88%E1%A8%B6%E1%A9%A0%E1%A8%B2%E1%A9%A5%E1%A8%B6%E1%A9%A0%E1%A8%B4%E1%A9%95%E1%A9%A5%E1%A8%BF#Pali) _santindriya_ 'tranquil', where the _ndri_ is of course to be encoded <NA, SAKOT, LOW TA, MEDIAL RA, SIGN I>.

**Comment by lianghai** [comment]
(I was just trying to catch up the discussion here and so far only I’ve gone through the last seven comments.)

@Richard57 and @mhosken:

> Secondly, in that proposed regex, 'V' must also include tone marks, as needed for ᨯ᩠ᩅ᩠᩶ᨿ /duai/ 'also, together' <LETTER DA, SAKOT, LETTER WA, TONE-2, SAKOT, LOW YA>.

Do you think there is attested contrast between “tone sign on the base” and “tone sign on the post-base spacing vowel sign”? (Can the latter happen for chained syllables?)

**Comment by Richard57** [comment]
I don't see the connection between the quotation and the question.

I don't believe there is any contrast between those two positions if you mean immediately on the post-base spacing vowel sign.  It's a different question if you mean after the consonants dependent on the post-base spacing vowel sign.

In extreme cases, albeit when phonetic rendering breaks down if I keep motion between grapheme clusters to the bare minimum, then the difference in placement before and after the sequence <SIGN AA, SAKOT, consonant> can matter, e.g. in words like the second word in the MFL dictionary,
ᨠ᩠ᩅᩬᩥᩣ᩠ᨶ᩠᩶ᨿ᩠ᨦ /kaː kin kuai kaːŋ kɔː/ 'the crows eat bananas in the middle of the grove', for which one has to resort to a graphical spelling - I used <HIGH KA, SAKOT, WA, SIGN OA BELOW, SIGN I, SIGN AA, SAKOT, NA, TONE-2, SAKOT, LOW YA, SAKOT, NGA>.  Patrick Chew has found a different spelling, with the tone mark on the first stack.  It also has NA and WA the other way round.

Most cases of syllables chained to SIGN AA seem to be orthographic CV, with subscripting originally being used to silence the vowel rather than to save palm leaf.  (Prepared palm leaf doesn't grow on trees, which is one reason it's been replaced by cardboard.)  I haven't yet found an example of <C1, SIGN AA, C1, MAI KANG> being abbreviated to <C1, SIGN AA, MAI KANG>, which would in general need a different rendering to the ordinary <C1, SIGN AA, MAI KANG>.

However, there remains the possibility of significant tone-marking, which is why I prefer to let the rendering system (engine + font) have all the information so it can sort it out.  I don't know what we'll find in China - Tai Lue seems to be fond of contracted syllables.

In OpenType terms, I don't like the idea of making SIGN AA a mark rather than a base.  There are constraints on when to give it a positive advance width so that one gets the same result with HarfBuzz and Microsoft engines.  The tone mark sequencing was set up to be graphically obvious while not inconsistent with the phonetic ordering principle. I think it will also request that the mechanism for skipping all but certain marks will have to skip sets of marks rather than all classes but one, though I might get away with reducing font robustness instead.  I wouldn't want the marks above to interact in *ᩈᩘᨠᩣ᩠ᨲᩥ <HIGH SA, MAI KANG LAI, HIGH KA, SIGN AA, SAKOT, HIGH TA, SIGN I>  I don't see any obvious reason why that hypothetical word shouldn't exist.  (The marks do interact in words like *<SA, MAI KANG LAI, HIGH KA, SIGN I>, which Lao texts do have, despite recommendations to write them with NGA instead of MAI KANG LAI.)

**Comment by mhosken** [comment]
ᨠ᩠ᩅᩬᩥᩣ᩠ᨶ᩠᩶ᨿ᩠ᨦ /kaː kin kuai kaːŋ kɔː/ is getting into calligraphy given that however it is spelled, it could never be analysed from its encoded form into its phonemic form. On that basis I would rule it out of scope for the USE. As to the hypothetical *ᩈᩘᨠᩣ᩠ᨲᩥ <HIGH SA, MAI KANG LAI, HIGH KA, SIGN AA, SAKOT, HIGH TA, SIGN I>, again I would suggest it will never occur because there is no precedent of having a tone over a SIGN AA in a non-calligraphic word.

**Comment by Richard57** [comment]
What has the behaviour of tone marks to do with the possible existence of the hypothetical *ᩈᩘᨠᩣ᩠ᨲᩥ <HIGH SA, MAI KANG LAI, HIGH KA, SIGN AA, SAKOT, HIGH TA, SIGN I>?  It has no tone marks!  The end of the word is exemplified by words like ᨻ᩠ᨿᩣ᩠ᨵᩥ /păɲaːt/ 'sickness'  <LOW PA, SAKOT, LOW YA, SIGN AA, SAKOT, LOW THA, SIGN I> (N3207 Section 14.8).

Now, MAI KANG LAI and SIGN I interact in words like ᩈᩘᩉᩥᨲ 'samhita' - for the style of the MFL, the interaction is that mai kang lai stays on the left consonant.  For Lao style fonts, the MAI KANG LAI moves to the following consonant and then the marks stack.  Now at present, I ensure the only the right marks interact by 'ignoring all marks except marks above'.  To use this technique on the rearranged marks in the hypothetical word, I have to make SIGN AA, now made a mark instead of being a base, into a 'mark above'.  Similar considerations applied to the sequence <MEDIA LA, SIGN AA, SAKOT, WA> need SIGN AA to be a mark below!  Now there is a solution  in OpenType - use sets of marks rather than a partition into classes, but everyone has to up their game - renderers, compilers and designers.

In short, marks following SIGN AA but attaching to the base are bad news.  On the other hand, the rightward movement of tone marks encoded before SIGN AA but attached to it if there are no intervening marks above is fairly straightforward.  I'm pretty sure I've seen tone marks non-contrastively positioned above SIGN AA in non-calligraphic writing.

**Comment by Richard57** [comment]
Thinking of apparently calligraphic forms, I have a related question on what I had taken to be the calligraphic form of the word for enlightenment, Pali _bodhi_.  It can be written in the form of a cross, with LOW PA in the middle, SIGN E on the left, SIGN AA on the right, subscript LOW THA below, and SIGN I above.  I have now come across it in _Pali_ in the accusative in plain (certainly not calligraphic!) text on p189 of Bunkhit Watcharasat's big blue book, as the final element of the word _mahābodhiṃ_.  Some element of graphic order is required for this word.  Now, for the nominative singular/stem form, the encoding <LOW PA, SAKOT, LOW THA, SIGN E, SIGN I, SIGN AA> is an obvious spelling and works with the USE, but I am tempted to simply swap SIGN I and SIGN AA in the phonetic order spelling as chained syllables.  But what do you recommend I do for the accusative?  Simply writing ᨻ᩠ᨵᩮᩥᩴᩣ <LOW PA, SAKOT, LOW THA, SIGN E, SIGN I, MAI KANG, SIGN AA> seems to work for some odd reason, although I thought MAI KANG was now classified as a vowel modifier, which would make 'sign am' work - but <SIGN AA, MAI KANG> doesn't render properly in an unresisting font.

**Comment by Richard57** [comment]
Just for clarification - 'sign am' is not appropriate for  _mahābodhiṃ_, even though it would work with some fonts' artistic designs.

**Comment by mhosken** [comment]
How do you want to spell it? Having said that I am working on a visual encoding order which will mess with that completely. But I do want to try to make something that supports the visual contrasts we need. To that end, I am looking for some contrasts:

- C1 1A6C 1A60 C2 contrasting (analagously) with C1 1A60 C2 1A6C, where C2 is not 1A3F.
- Visual contrast by position of U+1A7B (mai sam). Can we get away with one position in the grapheme cluster (where 1A63 starts a new cluster for the purposes of this question) for mai sam?

I'm hoping that we can come up with something that works for all needs even if it isn't a linguistic encoding that requires a typist to be able to read out the word they type, correctly.

(BTW sorry about confusing tone and vowel above, but I think I can support pya:ti.)

**Comment by Richard57** [comment]
Well, for cruciform _bodhi_ and the fuller forms of NT _khaokhɔːnɡ_ with one onset consonant, it would be nice to have a 'slide below consonant' alternative to SAKOT.  (For the latter, perhaps one could have a rule that SIGN AA only starts its own stack when one then meets a subscript consonant.)  I could then generate the accusative from the nominative by rule.  However, if we are going to use the renderer as a grammar checker, I would rather swap the last two vowels - so <LOW PA, SIGN E, SIGN I, MAI KANG, SAKOT, LOW THA, SIGN AA> or use CGJ as a sort of apostrophe to indicate an omitted (fictitious) consonant and use <LOW PA, SAKOT, LOW THA, SIGN E, SIGN I, MAI KANG, CGJ, SIGN AA>.

C1VC2 v. C1C2V for V=SIGN OA BELOW and C2 not being LO YA needs a long, hard search, because everyone seems to teach against the former.  ᩉᩬ᩠ᨾ ('good smell') v. ᩉ᩠ᨾᩬ ('learned expert') would be the obvious pair.  CVC does occur - Patrick Chew said either that they could or that they would be written the same, and the MFL entry for the 'hornet' words /tɔːtɛːn/ and /tɛːntɔː/ implies that they can be written the same.

As to MAI SAM, I'm not sure if the former is grammatical, but I think ᨧ᩠ᩃᩮᩥ᩠ᨿ᩻ /tɕălɤːj tɕălɤːj/ <HIGH CA, SAKOT, LA, SIGN E, SIGN I, SAKOT, LOW YA, MAI SAM> (reduplication without unambiguously marking the impurity of the onset) and ᨧ᩠ᩈ᩻ᩮᩥ᩠ᨿ /tɕălɤːj/ <HIGH CA, SAKOT, LA, MAI SAM, SIGN E, SIGN I, SAKOT, LOW YA> (no reduplication with impurity of onset doubly marked)  will be visually distinct in handwriting if <SAKOT, LA> and <SAKOT, LOW YA> do not overlap.

If you have the _Northern Thai Dictionary of Palm-Leaf Manuscripts_, you will find an example of such spacing in the word ᨧ᩠ᩃ᩠ᨿᨠ /tɕăliak/ 'to like' <HIGH CA, SAKOT, LA, SAKOT, LOW YA, HIGH KA> at the foot of p191.  However, the examples of visible differences do not rely on this feature.  The closest to a minimum pair I can find is ᨿᩥ᩠ᨷ᩻ /ɲip ɲip/ 'to blink' <LOW YA, SIGN I, SAKOT, BA, MAI SAM>, where MAI SAM is the rightmost superscript mark, and  ᨾᩱ᩶ᩈᩖᩥ᩻ᨯ /mai sălit/ 'a kind of tree' <MA, SIGN AI, TONE-2, HIGH SA, SAKOT, MEDIAL LA, MAI SAM, SIGN I, DA> where MAI SAM is the leftmost superscript mark of the middle cluster.  The relative positioning of vowel and MAI SAM does not seem to be consistent.  Pictures will be provided on request.

I believe the answer to the copy-typing issue is a spell-checker.  I have a colorisation solution for telling pronunciations apart - I've been using it for the Northern Thai Wikipedia.  (I use my Da Lekh Si font to see what I've typed, and to tell the spelling choices apart.)  The first _revision_ of the MFL dictionary loses any EU database protection some time this year.

I use '/' to transliterate SAKOT.

**Comment by mhosken** [comment]
@Richard57 could you email me. I'm having problems getting through to the email address I have for you. TIA.

**Comment by Richard57** [comment]
Mails getting through to me now - I didn't notice any problem.

**Comment by behdad** [comment]
Closing in preference to https://github.com/harfbuzz/harfbuzz/issues/2061
